storage/map_repository.hpp

```diff
@@ -0,0 +1,192 @@
+class MapRepository
+{
+public:
+  enum class ErrorCode;
+  using TNodeId = string;
+  using TOnSearchResultCallback = function<void (TNodeId const &)>;
+  using TOnStatusChangedCallback = function<void (TNodeId const &)>;
+  using TOnErrorCallback = function<void (TNodeId const &, ErrorCode)>;
+
+  /// \brief This enum describes status of a downloaded mwm or a group of downloaded mwm.
+  enum class ClientNodeStatus
+  {
+    UpToDate,              /**< Downloaded mwm(s) is up to date. No need update it. */
+    DownloadingInProcess,  /**< Downloading a new mwm or updating a old one. */
+    DownloadWasPaused,     /**< Downloading was paused or stopped by some reasons. E.g lost connection. */
+    NeedsToUpdate,         /**< An update for a downloaded mwm is ready according to county_attributes.txt. */
+    InQueue,               /**< A mwm is waiting for downloading in the queue. */
+  };
+
+  /// \brief Contains all properties for a node on the server.
+  struct ServerNodeAttrs
+  {
+    /// If it's not an extandable m_nodeSize is size of one mwm.
+    /// Otherwise m_nodeSize is a sum of all mwm sizes which belong to the group.
+    size_t m_nodeSize;
+    /// If the node is expandalbe (a big country) m_childrenCounter is number of children of this node.
+    /// If the node isn't expandable m_childrenCounter == -1.
+    int m_childrenCounter;
+    /// parentId is a node id of parent of the node.
+    /// If the node is "world" (that means the root) parentId == "".
+    TNodeId parentId;
+  };
+
+  /// \brief Contains all properties for a downloaded mwm.
+  /// It's applicable for expandable and not expandable node id.
+  struct ClientNodeAttrs
+  {
+    /// If it's not an extandable node m_nodeSize is size of one mwm.
+    /// Otherwise m_nodeSize is a sum of all mwm sizes which belong to the group and
+    /// have been downloaded.
+    size_t m_nodeSize;
+    /// If the node is expandable (a big country) m_mapsDownloaded is number of maps has been downloaded.
+    /// If the node isn't expandable m_mapsDownloaded == -1.
+    int m_mapsDownloaded;
+    /// \brief It's a mwm version which was taken for mwm header.
+    /// @TODO Discuss a version format. It should represent date and time (one second precision).
+    /// It should be converted easily to unix time.
+    /// \note It's set to zero in it's attributes of expandable node.
+    size_t m_mwmVersion;
+    /// A number for 0 to 100. It reflects downloading progress in case of
+    /// downloading and updating mwm.
+    uint8_t m_downloadingProgress;
+    ClientNodeStatus m_status;
+  };
+
+  /// \brief Error code of MapRepository.
+  enum class ErrorCode
+  {
+    NoError,                /**< An operation was finished without errors. */
+    NotEnoughSpace,         /**< No space on flash memory to download a file. */
+    NoInternetConnection,   /**< No internet connection. */
+    AnotherError,           /**< Wrong using of MapRepository or some unknown error. */
```

Comment:
Витя тебе писал про другое. Он говорил, что ошибки не имеют особого смысла для UI, а имеют смысл только для отладки. Т.е. если UI никак не должен реагировать на какие-то ошибки (т.е. их целиком и полностью игнорировать), то такой код ошибки не нужен - его можно выпилить и в коде вставить падение по CHECK().

> On Dec 11, 2015, at 13:35, Vladimir Byko-Ianko <notifications@github.com> wrote:
> 
> In storage/map_repository.hpp <https://github.com/mapsme/omim/pull/896#discussion_r47341481>:
> 
> > +    /// It should be converted easily to unix time.
> > +    /// \note It's set to zero in it's attributes of expandable node.
> > +    size_t m_mwmVersion;
> > +    /// A number for 0 to 100. It reflects downloading progress in case of
> > +    /// downloading and updating mwm.
> > +    uint8_t m_downloadingProgress;
> > +    ClientNodeStatus m_status;
> > +  };
> > +
> > +  /// \brief Error code of MapRepository.
> > +  enum class ErrorCode
> > +  {
> > +    NoError,                /**< An operation was finished without errors. */
> > +    NotEnoughSpace,         /**< No space on flash memory to download a file. */
> > +    NoInternetConnection,   /**< No internet connection. */
> > +    AnotherError,           /**< Wrong using of MapRepository or some unknown error. */
> AnotherError? Раньше были конкретные коды для конкрентых ошибок (в использовании загрузчика). Витя попросил их убрать и предложил сделать в этих ситуациях ASSERT-ты. Но, если будут сделаны ASSERT в дебаге, то в релизе надо что-то возвращать. Я не против того, чтоб вернуть коды конкретных ошибок. Договоритесь с Витей п-та :)
> 
> Кстати, как будет делаться проверка на 3G/Wifi? Ядром или отдельно в UI? Я имею ввиду диалог "вы точно уверены, что хотите загрузить 50Mb данных по 3G"?
> 
> Думаю, что в UI. (Отражу в диз доке)
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/mapsme/omim/pull/896/files#r47341481>.
>